src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonIEnumerableConverter.cs

```diff
@@ -0,0 +1,17 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// Base class for IEnumerable-based collections.
+    /// </summary>
+    internal abstract class JsonIEnumerableConverter<TCollection, TElement> : JsonResumableConverter<TCollection>
```

Comment:
do you imagine a new converter could derive from this to support (de)serializing `IAsyncEnumerable<T>`?

Reply:
The way I see [IAsyncEnumerable](https://github.com/dotnet/runtime/issues/1569) working is that the type that implements `IAsyncEnumerable` keeps the `DeserializationState` (aka `ReadStack`) and `JsonSerializationOptions` and then re-enters the serializer for each collection element, passing the `DeserializationState` and options. i.e. it by itself is not a converter -- it just forwards to whatever converter was present for the current particular collection.

Since the deserializer is built to support continuation with `DeserializationState` there shouldn't be much effort to get this functional.